src/containers/PerfherderGraphContainer/index.jsx

```diff
@@ -136,21 +135,31 @@ const styles = () => ({
 });
 
 class PerfherderGraphContainer extends Component {
-  state = {
-    data: null,
-    tooltipModel: null,
-    tooltipIsLocked: false,
-    canvas: null,
-    isLoading: false,
-  };
+  constructor(props) {
+    super(props);
+    const { timeDomain } = this.props;
+
+    if (missing(timeDomain)) {
+      Log.error('expecting a time range');
+    }
+
+    this.state = {
+      data: null,
+      tooltipModel: null,
+      tooltipIsLocked: false,
+      canvas: null,
+      isLoading: false,
+    };
+  }
 
   async componentDidMount() {
-    const { series, style, reference } = this.props;
+    const { series, style, reference, timeDomain } = this.props;
 
-    try {
-      this.setState({ isLoading: true });
+    this.setState({ isLoading: true });
+    await sleep(0); // ALLOW UI TO UPDATE
```

Comment:
Too much is happening synchronously!  When all the data is loaded (no network calls are made), a call to update the page takes a couple of seconds for a long calculation to run.  This locks up the UI for that time; preventing redrawing.  `await sleep(0)` adds a "yield" to the execution so that UI elements can update; so that `render()` is executed; before this long calculation.

Reply:
Instead of setting state in `componentDidMount`, you could do it in the constructor. This will make sure there is no extra render. This will also make sure the UI updates with the loading screen.

Reply:
https://reactjs.org/docs/react-component.html#componentdidmount

Reply:
I think this code belongs here: It has two main modes; the first time it is run (for the first instance of this component type) it takes a long time to run because lots of data is fetched from Treeherder.  So, I would like this component to show during that time; showing spinners in place of charts.  The subsequent runs (I guess for other instances of this same component), there is a good chance network requests are NOT made; and all promises resolve immediately. This method still takes long to run; not as long as the first run; but long enough to lock up the UI as the long promise chain selfishly resolves; and prevents other promise chains from executing; causing the lockup.  In this second case, I still want the ui to render spinners in place of the charts while we wait, but that will not happen unless I break up my promise chain to let others execute and actually let the UI update.  This single `sleep(0)` helps a lot, but more should be added so the ui is allowed to be responsive.